cmd/soroban-cli/src/commands/contract/deploy/wasm.rs

```diff
@@ -100,19 +108,96 @@ pub enum Error {
     Network(#[from] network::Error),
     #[error(transparent)]
     Wasm(#[from] wasm::Error),
+    #[error("cannot access config dir for alias file")]
+    CannotAccessConfigDir,
+    #[error(
+        "alias must be 1-30 chars long, and have only letters, numbers, underscores and dashes"
+    )]
+    InvalidAliasFormat { alias: String },
+    #[error(transparent)]
+    JsonSerialization(#[from] serde_json::Error),
+    #[error(transparent)]
+    Io(#[from] std::io::Error),
 }
 
 impl Cmd {
     pub async fn run(&self) -> Result<(), Error> {
+        self.validate_alias()?;
+
         let res = self.run_against_rpc_server(None, None).await?.to_envelope();
         match res {
             TxnEnvelopeResult::TxnEnvelope(tx) => println!("{}", tx.to_xdr_base64(Limits::none())?),
             TxnEnvelopeResult::Res(contract) => {
+                self.save_contract_id(&contract)?;
                 println!("{contract}");
             }
         }
         Ok(())
     }
+
+    fn validate_alias(&self) -> Result<(), Error> {
+        let alias = self.alias();
+
+        if alias.is_empty() {
+            return Ok(());
+        }
+
+        let regex = Regex::new(r"^[a-zA-Z0-9_-]{1,30}$").unwrap();
+
+        if regex.is_match(&alias) {
+            Ok(())
+        } else {
+            Err(Error::InvalidAliasFormat { alias })
+        }
+    }
+
+    fn alias(&self) -> String {
+        self.alias.clone().unwrap_or_default()
+    }
+
+    fn alias_path(&self) -> Result<PathBuf, Error> {
+        let config_dir = self.config.config_dir()?;
+        let network = self.config.network.network.clone().expect("must be set");
+        let alias = self.alias();
+        let file_name = format!("{alias}.json");
+
+        Ok(config_dir
+            .join("contract-ids")
+            .join(network)
+            .join(file_name))
+    }
+
+    fn save_contract_id(&self, contract: &String) -> Result<(), Error> {
+        if self.alias().is_empty() {
+            return Ok(());
+        }
```

Comment:
This fixed the bug, but not the root cause, that we're not leaning on the compiler to help us write safe logic with an Option<String>. So I would replace each of these if empty guards with a `if let Some(alias) = &self.alias { /* existing logic */ }`.

Reply:
Do you think the reliance on `self.alias()`, which ensures a string is always returned, is a anti-pattern in Rust? Should I drop it entirely so I can use `match` over the optional value in all places?

Reply:
I think it would be considered an anti-pattern. It's a more common approach in languages without a type like `Option<T>`. If an empty alias was a meaningful thing I'd say it was fine, in this case an empty alias is just another form of `Option::None`, so we're converting from one form to another form, and we're converting away from a form that the compiler is there to help protect us from those types of mistakes.

Reply:
I'd replace it with `if let Some(alias) = &self.alias {` or `match &self.alias {`.

Reply:
I would suggest. This acts as an assignment, but has the escape hatch.
```rust
    let Some(alias) = &self.alias else { return Ok(()); }
```

I'd also add alias as an argument to `alias_path` so
```rust
```rust
    let Some(alias) = &self.alias else { return Ok(()); }
    let file_path = self.alias_path(alias); // it's already a reference here
```
This way you don't need to worry about the existence of the alias in the other method.